File: src/Tex0tron/Tex0.cs
namespace Tex0tron;

public class Tex0
{
    public uint TextureBasePointer { get; set; }
    public int TextureBufferWidth { get; set; }
    public PixelStorageFormat TexturePixelFormat { get; set; }
    public int Width { get; set; }
    public int Height { get; set; }
    public bool HasAlpha { get; set; }
    public byte TextureFunction { get; set; }
    public uint ClutBufferBasePointer { get; set; }
    public PixelStorageFormat ClutPixelFormat { get; set; }
    public byte ClutStorageMode { get; set; }
    public int ClutEntryOffset { get; set; }
    public byte ClutBufferLoadControl { get; set; }

    public void Deserialize(ulong data)
    {
        TextureBasePointer = (ushort)(data & 0x3FFF) * 64u;
        TextureBufferWidth = (short)((data >> 14) & 0x3F) * 64;
        TexturePixelFormat = (PixelStorageFormat)((data >> 20) & 0x3F);
        Width = 1 << (short)((data >> 26) & 0xF);
        Height = 1 << (short)((data >> 30) & 0xF);
        HasAlpha = ((data >> 34) & 0x1) == 1;
        TextureFunction = (byte)((data >> 35) & 0x3);
        ClutBufferBasePointer = (ushort)((data >> 37) & 0x3FFF) * 64u;
        ClutPixelFormat = (PixelStorageFormat)((data >> 51) & 0xF);
        ClutStorageMode = (byte)((data >> 55) & 0x1);
        ClutEntryOffset = (byte)((data >> 56) & 0x1F) * 16;
        ClutBufferLoadControl = (byte)((data >> 61) & 0x7);
    }
}

File: src/Tex0tron.Web/src/app/tex0/tex0.ts
export enum PixelStorageFormat {
  PsmCt32 = 0x00,
  PsmCt24 = 0x01,
  PsmCt16 = 0x02,
  PcmCt16S = 0x0A,
  PsmT8 = 0x13,
  PsmT4 = 0x14,
  PsmT8H = 0x1B,
  PsmT4HL = 0x24,
  PsmT4HH = 0x2C,
  PsmZ32 = 0x30,
  PsmZ24 = 0x31,
  PsmZ16 = 0x32,
  PsmZ16S = 0x3A
}

export interface Tex0 {
  textureBasePointer: number;
  textureBufferWidth: number;
  texturePixelFormat: PixelStorageFormat;
  width: number;
  height: number;
  hasAlpha: boolean;
  textureFunction: number;
  clutBufferBasePointer: number;
  clutPixelFormat: PixelStorageFormat;
  clutStorageMode: number;
  clutEntryOffset: number;
  clutBufferLoadControl: number;
}
